analyzer/src/main/kotlin/managers/SpdxDocumentFile.kt

```diff
@@ -100,6 +103,16 @@ internal fun SpdxDocument.projectPackage(): SpdxPackage? =
         // directory itself).
         ?.singleOrNull { it.packageFilename.isEmpty() || it.packageFilename == "." }
 
+/**
+ * Try to find an [SpdxExternalReference] in this [SpdxPackage] of type purl from which the scope of a
+ * package manager dependency can be extracted. Return this scope or *null* if cannot be determined.
+ */
+internal fun SpdxPackage.extractScopeFromExternalReferences(): String? =
+    externalRefs.filter { it.referenceType == SpdxExternalReference.Type.Purl }
+        .mapNotNull { REG_SCOPE.matchEntire(it.referenceLocator) }
+        .map { it.groupValues[1] }
+        .firstOrNull()
```

Comment:
I propose to instead implement this by creating a `URI` from the PURL and use our `URI.getQueryParameters()` extension function.

Reply:
There's also the `String.toUri` extension function which should make this rather readable, like

    externalRefs.filter { it.referenceType == SpdxExternalReference.Type.Purl }
        .mapNotNull { it.referenceLocator.toUri { uri -> uri.getQueryParameters() }.getOrNull() }
        .firstNotNullOfOrNull { it["scope"]?.singleOrNull() }

Reply:
I tried this, but found an unexpected problem with this approach: The PURL to reference the definition file looks like this: `pkg:conan/test.org@1.2.3?scope=someScope`. According to the [Javadocs of the URI class](https://docs.oracle.com/javase/8/docs/api/java/net/URI.html), this is not a hierarchical URI because the scheme-specific part does not start with a slash. Therefore, this part is not further parsed, and the query string is not processed. For this reason, the `getQueryParameters()` function returns an empty map.

Reply:
Crap, that looks like a design-flaw of PURLs to me... anyway, I'd probably just add another mapping then:

    externalRefs.filter { it.referenceType == SpdxExternalReference.Type.Purl }
        .map { it.referenceLocator.replace("pkg:", "pkg://") }
        .mapNotNull { it.toUri { uri -> uri.getQueryParameters() }.getOrNull() }
        .firstNotNullOfOrNull { it["scope"]?.singleOrNull() }

Reply:
Changed to something close to your proposal.